feat: full PayRef implementation

fluffypony opened the pull request:
## Description
---

Implements the Payment Reference (PayRef) system for Tari, providing globally unique identifiers for individual transaction outputs to enable payment verification in Mimblewimble-based blockchains. PayRefs solve the critical business problem of payment verification for exchanges and merchants without compromising privacy.

**Key Features:**
- Generate PayRefs using `PayRef = Blake2b_256(block_hash || output_hash)` for merkle proof compatibility
- Transaction-based PayRef calculation ensuring sender and receiver generate identical references
- Support for multiple PayRefs per transaction (multi-output payments)
- Confirmation-based stability with reorg safety (default: 5 confirmations)
- Privacy-preserving payment verification
- Full FFI support for external integrations
- Exchange documentation and integration examples

## Motivation and Context
---

The primary use case is when users send payments to exchanges but the exchange claims they haven't received the payment. Currently, there's no reliable way to prove payments in Mimblewimble due to:

- **Kernel obfuscation**: No direct link between kernels and specific outputs
- **Receiver limitations**: Receivers never see kernels in one-sided payments  
- **Privacy by design**: Transaction relationships are intentionally hidden

This creates significant business friction:
- High customer support burden for exchanges
- Users cannot prove payments were sent
- Difficulty maintaining audit trails
- Adoption friction due to payment verification problems

PayRefs provide a practical solution that ensures both sender and receiver can generate the same payment reference for verification, while maintaining privacy and enabling future merkle proof verification against block headers.

## How Has This Been Tested?
---

- **Unit tests**: PayRef generation with new formula, multiple output scenarios, change output identification
- **Integration tests**: End-to-end payment flows ensuring sender-receiver PayRef matching
- **Reorg tests**: Verification of PayRef cleanup during blockchain reorganizations
- **Manual testing**: Console wallet transaction history display with multiple PayRefs support
- **API testing**: gRPC endpoints for PayRef retrieval supporting multiple references per transaction
- **Edge case validation**: Handling of spent outputs, burned outputs, and same-block spends

## What process can a PR reviewer use to test or verify this change?
---

**Basic PayRef Generation:**
1. Send a transaction with multiple outputs using the console wallet
2. Wait for 5+ confirmations 
3. Check transaction history - PayRefs should appear as "Available"
4. Verify each PayRef is a 64-character hex string
5. Confirm receiver generates identical PayRefs for the same outputs

**Multi-Output Transactions:**
1. Create a transaction with multiple recipients
2. Verify multiple PayRefs are displayed (one per non-change output)
3. Confirm each recipient can verify their specific PayRef

**PayRef Search:**
1. In console wallet, press `t` for transactions, then `s` to search
2. Enter a known PayRef (full or partial)
3. Verify it finds the matching transaction

**Reorg Safety Testing:**
1. Force a reorg on a test network
2. Verify PayRefs are properly cleaned up for reorged transactions
3. Confirm no orphaned PayRef entries remain

**FFI Interface:**
1. Build with `cargo build`
2. Check that new FFI functions compile without errors
3. Verify header file `wallet.h` includes new PayRef structures with array support

**Documentation:**
1. Review exchange integration guide in `docs/src/09_adding_tari_to_your_exchange.md`
2. Check user guide in `docs/src/payref_userguide.md`
3. Verify code examples reflect the new multi-PayRef support

**API Testing:**
```bash
# Test PayRef retrieval for specific transaction
grpcurl -plaintext -d '{"tx_id": "YOUR_TX_ID"}' \
  localhost:18143 tari.rpc.Wallet/GetTransactionPayRefs

# Test payment lookup by PayRef  
grpcurl -plaintext -d '{"payment_reference_hex": "YOUR_PAYREF_HERE"}' \
  localhost:18143 tari.rpc.Wallet/GetPaymentByReference

# Verify multiple PayRefs for multi-output transactions
grpcurl -plaintext localhost:18143 tari.rpc.Wallet/GetCompletedTransactions
```

## Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork  
- [ ] Other - Please specify

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced Payment Reference (PayRef) support across Tari blockchain outputs, enabling unique, privacy-preserving identifiers.
  - Added CLI commands and UI enhancements in the console wallet for displaying, searching, and verifying PayRefs.
  - Enabled gRPC and FFI APIs for retrieving and verifying transactions by PayRef, facilitating integration with external services.
  - Implemented database schema migrations, indexing, and backend support for efficient PayRef lookups.
  - Added a BaseNode gRPC streaming method to search blockchain outputs by payment references.
  - Extended wallet and transaction services with PayRef generation, storage, and querying capabilities.
  - Added PayRef verification and display configuration options in the wallet UI.

- **Documentation**
  - Added comprehensive user guides and integration documentation for PayRef usage and best practices.
  - Updated application and API overviews to include PayRef features.
  - Added detailed gRPC and FFI overviews covering PayRef functionality.

- **Bug Fixes / Refactor**
  - Improved error handling and code clarity across multiple modules.
  - Streamlined output hash tracking and transaction model updates to support PayRef features.
  - Simplified error constructions and iterator usage in core and communication modules for cleaner code.

- **Chores**
  - Upgraded Diesel dependency in multiple crates for improved compatibility and security.
  - Added new hash domain for Payment Reference generation.
  - Minor code cleanups and simplifications across core and comms crates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fluffypony committed:
initial commit

fluffypony committed:
add payref to output manager

fluffypony committed:
implement in gRPC

fluffypony committed:
add to existing gRPC outputs

fluffypony committed:
change how we get the block hash

fluffypony committed:
update gRPC responses

fluffypony committed:
return hex instead of bytes for payref

fluffypony committed:
fix GetCompletedTransactions to return payref

fluffypony committed:
add payref commands to console wallet

fluffypony committed:
fix payref in TUI

fluffypony committed:
FFI payref integration

fluffypony committed:
add payref searching to the base node

fluffypony committed:
add payref indexing so the base node doesn't get chewed up

fluffypony committed:
move payref to its own crate, add docs and test

fluffypony committed:
fix errors

fluffypony committed:
build PayRef index on first start after upgrade

stringhandler committed:
Merge branch 'development' into payref

stringhandler reviewed (dismissed):
utack

fluffypony committed:
fix clippy complaints + test compiling

fluffypony committed:
Merge branch 'payref' of https://github.com/fluffypony/tari into payref

fluffypony committed:
fix incorrect FFI implementation

fluffypony committed:
fix coderabbit issues

fluffypony committed:
fix build + linting

fluffypony committed:
unify config

fluffypony committed:
fix duplicate function

fluffypony committed:
fix coderabbit nitpicks

fluffypony committed:
remove pagination + dupes

fluffypony committed:
use crate constant instead of hardcoded value

fluffypony committed:
clean up gRPC by removing pagination

SWvheerden committed:
Merge branch 'development' into payref

SWvheerden commented on applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs:
```diff
@@ -202,6 +213,110 @@ impl WalletGrpcServer {
+    let mut payment_refs = Vec::new();
+    
+    // Note: Don't skip based on transaction mined_height for restored wallets
+    // Individual outputs may have mining data even if transaction doesn't
```
This should not be the case, there was a bug that was recently fix where this did occur. The mined height in the transaction and mined_height of the output should be in sync

SWvheerden commented on applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs:
```diff
@@ -202,6 +213,110 @@ impl WalletGrpcServer {
+            warn!(target: LOG_TARGET, "Failed to get spent outputs for PayRef calculation: {}", e);
+            Vec::new()
+        }
+    };
```
This will not be scalable, the internal DB in the OMS has methods to ask for specific outputs, we just need to expose those requests. We don't need to pull out the entire DB to memory here.

SWvheerden commented on applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs:
```diff
@@ -1103,6 +1223,22 @@ impl wallet_server::Wallet for WalletGrpcServer {
+                    .unwrap_or_default()
+                    .iter()
+                    .map(|b| format!("{:02x}", b))
+                    .collect::<String>();
```
this is wrong in two ways:
1- if there is no payref we should make it none, and not the default. 
2- We should not return the first payref,most transactions have 2 outputs, 1 for change, and 1 for sent. We need to include both or only send the sent output one. 

I would only do the sent out, and do figure out which one is the sent one, we need to search the oms for the one that does not exist.

SWvheerden commented on applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs:
```diff
@@ -1200,55 +1337,57 @@ impl wallet_server::Wallet for WalletGrpcServer {
+                .collect();
+            
+            // Calculate PayRef for this transaction's first output (if mined)
+            let payment_refs = self.calculate_payment_references_for_transaction(&output_commitments, txn.mined_height.unwrap_or(0)).await;
```
See above, we cant use first output. 
Also if the txn.mined_height is none, or = 0, we can assume its not mined.

SWvheerden commented on applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs:
```diff
@@ -1283,57 +1422,61 @@ impl wallet_server::Wallet for WalletGrpcServer {
+                .collect();
+            
+            // Calculate PayRef for this transaction's first output (if mined)
+            let payment_refs = self.calculate_payment_references_for_transaction(&output_commitments, txn.mined_height.unwrap_or(0)).await;
```
same as above

SWvheerden commented on applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs:
```diff
@@ -1620,7 +1763,190 @@ impl wallet_server::Wallet for WalletGrpcServer {
+
+        let mut output_service = self.get_output_manager_service();
+        
+        match output_service.find_payment_by_reference(payment_ref).await {
```
This will be very limited, you will only be able to search your own received outputs on payment ref, not sent. 
We need to implement this on the TMS and per sent transaction rather,

SWvheerden commented on applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs:
```diff
@@ -202,6 +213,110 @@ impl WalletGrpcServer {
     }
+
+    /// Calculate PayRefs for a transaction's outputs using existing wallet output data
+    async fn calculate_payment_references_for_transaction(
+        &self,
+        output_commitments: &[Vec<u8>],
+        mined_height: u64,
+    ) -> Vec<Vec<u8>> {
+        calculate_payment_references_impl(
+            self.get_output_manager_service(),
+            output_commitments,
+            mined_height,
```
```suggestion
    async fn calculate_payment_references_for_transaction(
        output_manager: &OutputManagerHandle,
        output_commitments: &[Vec<u8>],
        mined_height: u64,
    ) -> Vec<Vec<u8>> {
        calculate_payment_references_impl(
            output_manager,
            output_commitments,
            mined_height,
```
By doing this, you don't have to pass around the grpc wallet service

SWvheerden commented on applications/minotari_console_wallet/src/ui/state/app_state.rs:
```diff
@@ -777,10 +777,99 @@ impl AppStateInner {
         Ok(())
     }
 
+    async fn calculate_payment_references_for_transactions(&mut self) -> Result<(), UiError> {
```
This is def going to be displaying different results in different wallets.

SWvheerden reviewed (changes requested):
I believe there are some fundamental problems in the way this is implemented. 
There are some improvements that need to be made for this to be scalable such as don't query the entire db if you want a single commitment etc. 

But the issue here is in how transactions works. 
When a transaction is created the TMS saves this transaction, the OMS will only save your outputs. So not the sent outputs. So a sender can never create a payref from the OMS as the sent outputs are not stored in there. 

The other issue here is that a tx has on avg 2 outputs, maybe less maybe more. One is a change coming back to you if there is a change, other is for each output going out. 

What we need to do is create a payref from the TMS when we create the tx, it might need to be stored in an incomplete state, so that you can calculate the payref when you want for the tx. 
But the sender when creating the tx will know which tx is its own change, and which is the outputs going out. We need to flag those outputs so we can create payrefs from them. 
We can then just ask the TMS for the payref for a tx, by looking at the sent outputs, and taking the mined_height and calculating the payrefs.

fluffypony committed:
Update PayRef formula from commitment to output_hash

- Change PayRef = Blake2b_256(block_hash || commitment) to Blake2b_256(block_hash || output_hash)
- Update imports to use HashOutput instead of CompressedCommitment
- Update documentation and examples to reflect new formula
- This enables better security and performance for payment verification

fluffypony committed:
Add output hash storage to transaction models

- Add HashOutput import to transaction storage models
- Add sent_output_hashes, received_output_hashes, and change_output_hashes fields
- Update constructors to initialize new fields with empty vectors
- Update all conversion functions to properly handle output hash fields
- This enables fast PayRef generation without scanning full transaction outputs

fluffypony committed:
Implement fast PayRef calculation using stored output hashes

- Update output manager to use output_hash instead of commitment for PayRef generation
- Create new payref_calculator module for efficient transaction PayRef calculations
- Add calculate_transaction_payrefs function for super fast PayRef generation
- Add find_payment_by_reference function for optimized PayRef lookup
- Add comprehensive tests for new PayRef calculation functionality
- This enables O(1) PayRef generation using pre-computed output hashes

fluffypony committed:
Fix compilation errors in PayRef implementation

- Fix PaymentDirection enum values (use Sent/Received instead of Outbound/Inbound)
- Fix PaymentDetails struct construction to match actual fields
- Add missing output hash fields to SQLite database transaction constructors
- Fix payment_id method call (use to_bytes() instead of as_bytes())
- Remove unused ByteArray import
- Add TODO comments for future database schema updates

fluffypony committed:
Update core blockchain database to use output_hash for PayRef generation

- Update LMDB database to use output_hash instead of commitment for PayRef generation
- Fix generate_payment_reference_for_output function signature and all callsites
- Update PayRef calculation throughout chain storage to use new formula
- Fix TxId type error in payref_calculator test
- This ensures consistency between wallet and blockchain PayRef generation

fluffypony committed:
Add wallet ownership detection methods to OutputManager

fluffypony committed:
Complete OutputManager handle interface for PayRef integration

fluffypony committed:
Implement receiver protocol output hash capture for PayRef

fluffypony committed:
Add service method implementations for PayRef API with compilation errors to fix

fluffypony committed:
Fix PayRef service method implementations to compile successfully

fluffypony committed:
Add FFI API functions for PayRef functionality

fluffypony committed:
Add gRPC API definitions for PayRef functionality

fluffypony committed:
Enhance console wallet PayRef commands to use new multiple PayRef APIs

fluffypony committed:
Fix hex encoding usage in PayRef commands

fluffypony committed:
Implement missing gRPC methods and fix compilation issues

SWvheerden commented on base_layer/core/src/base_node/comms_interface/comms_request.rs:
```diff
@@ -83,6 +83,8 @@ pub enum NodeCommsRequest {
     FetchUnspentUtxosInBlock {
         block_hash: BlockHash,
     },
+    FetchOutputByPayRef([u8; 32]),
```
```suggestion
    FetchOutputByPayRef(FixedHash),
```
we use 32byte hashes all over and parse this as fixedhash so that we don't have to handle conversions everywhere

SWvheerden commented on base_layer/core/src/chain_storage/async_db.rs:
```diff
@@ -433,3 +435,23 @@ mod tests {
+        tokio::task::spawn_blocking(move || {
+            let backend = db.db_read_access()?;
+            backend.check_output_spent_status(output_hash)
+        }).await.map_err(|_| ChainStorageError::AccessError("Task join error".to_string()))?
```
why not use the above macro? `make_async_fn!`

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -56,6 +59,7 @@ use tari_utilities::{
     hex::{to_hex, Hex},
     ByteArray,
 };
+// PayRef generation moved to shared utility crate
```
```suggestion
```

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -30,6 +30,8 @@ use std::{
     time::Instant,
 };
 
+// Blake2b and digest removed - using shared PayRef utility
```
```suggestion
```

fluffypony committed:
Add PayRef columns to database schema and SQL structs

fluffypony committed:
Fix all struct initializations with PayRef fields

fluffypony committed:
Add basic PayRef unit tests and start fixing test structs

fluffypony committed:
Add basic PayRef integration tests - core functionality works

fluffypony committed:
Implement PayRef serialization/deserialization helpers and fix first test struct

fluffypony committed:
Fix PayRef serialization/deserialization in transaction storage

fluffypony committed:
Fix remaining test CompletedTransaction struct initializations

fluffypony committed:
Fix PayRef error handling types in serialization

fluffypony committed:
Implement PayRef output hash population when transactions are mined

fluffypony committed:
Implement gRPC PayRef methods using actual transaction data

fluffypony committed:
Fix gRPC PayRef compilation errors

fluffypony committed:
Fix PayRef test types and imports (partial)

fluffypony committed:
Add PayRef cleanup during block deletion for reorgs

fluffypony committed:
Complete PayRef reorg cleanup implementation - all tests passing

fluffypony committed:
fix coderabbit issue

fluffypony committed:
add a space to the TUI description

fluffypony committed:
Add PayRef migration batch commit optimization

fluffypony committed:
Add PayRef migration checkpointing support

fluffypony committed:
Add PayRef migration error handling and recovery

fluffypony committed:
Replace PayRef cleanup error suppression with proper handling

fluffypony committed:
Add transaction verification and replace panic-prone code

fluffypony committed:
Update PayRef service documentation for consistency

fluffypony committed:
Enhance FFI documentation and clarify response types

fluffypony committed:
Fix critical PayRef generation bug - use proper Blake2b hash

fluffypony committed:
Enforce minimum confirmation requirements for PayRef generation

fluffypony committed:
Improve PayRef migration error handling and resource monitoring

fluffypony committed:
Add detailed documentation for wallet_generate_payment_reference_from_data

fluffypony committed:
Add detailed documentation for wallet_format_payment_reference

fluffypony committed:
Fix compilation errors by using correct error variant

fluffypony committed:
Fix compilation errors in PayRef migration code

fluffypony committed:
Improve PayRef cleanup error handling in delete_block_inputs_outputs

fluffypony committed:
Add granular error handling and resource monitoring to PayRef migration batch processing

fluffypony committed:
Add database integrity verification before PayRef migration

fluffypony committed:
feat: add docs with payref included

fluffypony committed:
Increase PayRef migration disk space thresholds: 50MB->500MB warn, 200MB critical abort

SWvheerden commented on applications/minotari_console_wallet/src/ui/state/app_state.rs:
```diff
@@ -777,10 +777,99 @@ impl AppStateInner {
+        };
+        
+        // Process all outputs (unspent and spent) to build the lookup map
+        for output in unspent_outputs.into_iter().chain(spent_outputs.into_iter()) {
```
calculate it directly from the tx.

fluffypony committed:
Add batch commitment query to OutputManager for PayRef optimization

fluffypony committed:
Optimize PayRef lookup using targeted database query

fluffypony committed:
Add multiple payment_references support to TransactionInfo

fluffypony committed:
Fix multiple PayRef return bug - populate all PayRefs instead of just first

fluffypony committed:
Optimize UI PayRef calculation using batch commitment query

fluffypony committed:
Add database indexes for PayRef performance optimization

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -748,7 +889,36 @@ impl LMDBDatabase {
+                },
+                Err(e) => {
+                    // Actual database error - log warning but don't fail the transaction
+                    warn!(target: LOG_TARGET, "Failed to delete PayRef for output {}: {}", output_hash.to_hex(), e);
```
on this we need to error, and let the node retry the tx , or what ever higher end logic there is. 
if we allow this, we cause the db to be in a broken state

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -1038,6 +1208,29 @@ impl LMDBDatabase {
+                    Err(ChainStorageError::ValueNotFound { .. }) => {
+                        // Expected case for outputs that didn't have PayRefs
+                    },
+                    Err(e) => warn!(target: LOG_TARGET, "Failed to delete PayRef during reorg for output {}: {}", output_hash.to_hex(), e),
```
should error

SWvheerden commented on base_layer/wallet/src/output_manager_service/storage/models.rs:
```diff
@@ -87,6 +89,104 @@ impl DbWalletOutput {
+                    PaymentDirection::SentChange
+                } else {
+                    PaymentDirection::Sent
+                }
```
```suggestion
if self.payment_id == PaymetID::TransactionInfo{
PaymentDirection::SentChange}else {PaymentDirection::Sent}
```
This is not maybe, this is guaranteed to be accurate.

SWvheerden commented on base_layer/wallet/src/transaction_service/handle.rs:
```diff
@@ -208,6 +208,16 @@ pub enum TransactionServiceRequest {
+    /// Get all PayRefs for a specific transaction
+    GetTransactionPayRefs(TxId),
+    /// Get transaction details for a PayRef (enhanced with multiple recipients)
+    GetPaymentByReference([u8; 32]),
```
```suggestion
    GetPaymentByReference(FixedHash),
```

SWvheerden commented on base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs:
```diff
@@ -986,6 +993,50 @@ where
+                        Err(e) => {
+                            warn!(target: LOG_TARGET, "Error checking if output is change: {:?}", e);
+                            // Default to treating as sent to be safe
+                            sent_hashes.push(output_hash);
```
I think we should error here, something is wrong, don't save wrong results into the database

SWvheerden commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -3911,6 +3923,185 @@ where
     }
+
+    /// Get PayRefs for a specific transaction with verification
+    fn get_transaction_payrefs(&self, tx_id: TxId) -> Result<Vec<[u8; 32]>, TransactionServiceError> {
```
```suggestion
    fn get_transaction_payrefs(&self, tx_id: TxId) -> Result<Vec<Fixedhash>, TransactionServiceError> {
```

SWvheerden commented on base_layer/wallet/src/transaction_service/storage/sqlite_db.rs:
```diff
@@ -68,6 +68,20 @@ use crate::{
+        None => Ok(Vec::new()),
+    }
+}
+
```
We dont need to use bincode here, its already in raw byte form

SWvheerden commented on base_layer/wallet/src/transaction_service/storage/sqlite_db.rs:
```diff
@@ -1409,6 +1426,7 @@ impl InboundTransactionSql {
             last_send_timestamp: i.last_send_timestamp.map(|t| t.naive_utc()),
             payment_id: Some(i.payment_id.to_bytes()),
             user_payment_id,
+            received_output_hashes: Some(serialize_output_hashes(&i.received_output_hashes)?),
```
```suggestion
            received_output_hashes: Some(i.received_output_hashes.to_vec()),
```

SWvheerden commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -231,6 +236,49 @@ pub type TariComAndPubSignature = ComAndPubSignature;
 pub type TariUnblindedOutput = UnblindedOutput;
 pub struct TariUnblindedOutputs(Vec<UnblindedOutput>);
 
+/// PayRef Configuration FFI Types
```
Can we move all of this to a separate file we did with `ffi_basenode_state`

SWvheerden replied:
We can also add from method to convert from the tari core methods to these c structs

fluffypony committed:
Fix compilation errors in PayRef implementation

SWvheerden reviewed (changes requested):
Looking good, 
few comments I think we need to address
and I think rather than use `[u8;32]` everywhere for the payref, we should simply replace that with `Fixedhash`  as that already wraps that and then we get all the helper methods and type safety

fluffypony committed:
Fix remaining compilation errors and warnings

fluffypony committed:
Fix clippy::io_other_error lint issue

fluffypony committed:
Remove deprecated payment_reference field from gRPC schema

fluffypony committed:
Replace [u8; 32] with FixedHash in FetchOutputByPayRef

Credit: @SWvheerden for pointing out this type safety improvement

fluffypony committed:
Replace manual async implementation with make_async_fn! macro

Credit: @SWvheerden for suggesting this cleaner approach

fluffypony committed:
Remove outdated comments from lmdb_db.rs

Credit: @SWvheerden for noting these are no longer relevant

fluffypony committed:
Change PayRef deletion warnings to errors

Ensure database inconsistencies fail transactions instead of just logging warnings.

Credit: @SWvheerden for pointing out this important error handling improvement

fluffypony committed:
Replace [u8; 32] with FixedHash in GetPaymentByReference

Credit: @SWvheerden for highlighting this type safety improvement

fluffypony committed:
Error instead of defaulting when output change check fails

Prevents silent data corruption by returning proper errors instead of defaulting.

Credit: @SWvheerden for pointing out this important reliability improvement

fluffypony committed:
Change get_transaction_payrefs return type to Vec<FixedHash>

Credit: @SWvheerden for pointing out this type safety improvement

fluffypony committed:
Remove bincode serialization for output hashes, use raw byte form

Replace serialize/deserialize_output_hashes functions with direct Vec operations for better performance and simplicity.

Credit: @SWvheerden for suggesting this optimization

fluffypony committed:
Fix FixedHash type compatibility issues in wallet service

- Update get_payment_by_reference to accept FixedHash
- Fix TransactionPayRefs response type
- Update protocol error handling
- Convert PayRef arrays to FixedHash for type consistency

Credit: @SWvheerden for highlighting these type safety improvements

fluffypony committed:
Fix console wallet PayRef calculation using direct transaction approach

fluffypony committed:
Simplify wallet output direction logic using PaymentId check

fluffypony committed:
Change SQLite schema to use Vec<FixedHash> directly

fluffypony committed:
Add missing FixedHash imports to async_db.rs and comms_request.rs

fluffypony committed:
Add missing BlockchainDatabase method implementations

fluffypony committed:
Remove unused doc comments from async macros

fluffypony committed:
Fix core compilation errors and type consistency issues

fluffypony committed:
Revert SQLite schema changes and implement proper FixedHash conversion

fluffypony committed:
Update transaction models to use Vec<FixedHash> and fix type consistency

fluffypony committed:
Fix all remaining compilation errors and type mismatches

fluffypony committed:
Fix clippy error by changing loop to conditional in PayRef generation

fluffypony committed:
Refactor duplicate payment reference functions with helper functions

fluffypony committed:
Fix formatting logic duplication with TODO for custom format

fluffypony committed:
Add standardized documentation to PayRef functions in wallet.h

fluffypony committed:
Add bounds checking for PayRef hex string slicing to prevent panic

fluffypony committed:
Fix compilation errors in refactored payment reference functions

fluffypony committed:
Apply code formatting fixes for final linting compliance

fluffypony committed:
Add OutputWithPayRef data structures for per-output PayRef tracking

fluffypony committed:
Update service.rs with per-output PayRef generation logic

fluffypony committed:
Update gRPC server to use per-output PayRef approach

fluffypony committed:
Update console wallet UI to display per-output PayRefs

fluffypony committed:
Add FFI interface for per-output PayRef structures and functions

fluffypony committed:
Add tests for per-output PayRef structures

fluffypony committed:
Fix compilation errors in gRPC and FFI interfaces

fluffypony committed:
Import DbWalletOutput for PayRef output data

fluffypony committed:
Add output manager get_output_by_hash method for PayRef

fluffypony committed:
Replace mock data with real output amounts and status from output manager

fluffypony committed:
Implement wallet_get_payref_for_output_hash FFI function

fluffypony committed:
Fix compilation errors: remove invalid OutputStatus variant and unused import

fluffypony committed:
Fix clippy warnings: box large enum variant and remove problematic test

fluffypony committed:
Apply clippy fixes and remove test module reference

fluffypony commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -9469,136 +9521,237 @@ pub unsafe extern "C" fn wallet_cancel_pending_transaction(
+            ptr::null_mut()
         },
     }
 }
```
The FFI doesn't have a database...?

fluffypony committed:
Remove redundant conversion in output_hashes processing

fluffypony committed:
Update core PaymentReference type from [u8; 32] to FixedHash

fluffypony committed:
Update PaymentDetails and PaymentRecord structs to use FixedHash

fluffypony committed:
Update transaction service structs to use FixedHash

fluffypony committed:
Update output manager handle enum and method to use FixedHash

fluffypony committed:
Update blockchain backend interface to use FixedHash

fluffypony committed:
Update LMDB database methods to use FixedHash

fluffypony committed:
Update blockchain database method to use FixedHash

fluffypony committed:
Update async database macros to use FixedHash

fluffypony committed:
Update local interface methods to use FixedHash

fluffypony committed:
Update test helper methods to use FixedHash

fluffypony committed:
Update output manager service methods to use FixedHash

fluffypony committed:
Update FFI layer to use FixedHash

fluffypony committed:
Fix inbound handler PayRef usage for FixedHash

fluffypony committed:
Fix compilation errors for FixedHash conversion

fluffypony committed:
Fix remaining FFI compilation errors for FixedHash

fluffypony committed:
Fix clippy warnings and test compilation error

fluffypony committed:
Optimize console wallet PayRef calculation to use transaction service

fluffypony committed:
Optimize OMS PayRef methods to use transaction service instead of output scanning

fluffypony committed:
Fix compilation errors and unused variable warnings

fluffypony committed:
Clean up unused variables and mark deprecated methods as allow(dead_code)

SWvheerden committed:
format

SWvheerden committed:
wip cleanup

SWvheerden committed:
some changes

SWvheerden committed:
change command

SWvheerden committed:
more stuff

SWvheerden committed:
more work

SWvheerden committed:
more wip

SWvheerden committed:
done wallet grpc server

SWvheerden committed:
more delete

SWvheerden committed:
more work

SWvheerden committed:
more cleanup

SWvheerden committed:
cleanup

SWvheerden committed:
cleanup

SWvheerden committed:
more cleanup

SWvheerden committed:
fixes

SWvheerden committed:
more cleanup

SWvheerden committed:
fix trp

SWvheerden committed:
cleanup stp

SWvheerden committed:
cleanup ffi

SWvheerden committed:
final cleanup ffi

SWvheerden committed:
cleanup

SWvheerden committed:
More cleanup

stringhandler commented on applications/minotari_app_grpc/proto/base_node.proto:
```diff
@@ -559,3 +561,48 @@ message LivenessResult{
+
+
+// Status of payment reference lookup
+enum PaymentReferenceStatus {
```
This doesn't seem to be used

SWvheerden committed:
fix

SWvheerden committed:
more removed

SWvheerden committed:
more fixes

SWvheerden committed:
oms models

SWvheerden committed:
build

SWvheerden committed:
More updates

SWvheerden committed:
exstra debug

SWvheerden committed:
Merge branch 'development' into payref

SWvheerden committed:
fmt

SWvheerden committed:
fix migrate

Commit:
Fix payment id cutoff

Fixed payment id cutoff display on the console wallet detail transactions tab

SWvheerden committed:
fix for utxo

SWvheerden committed:
fix db migration bug

SWvheerden committed:
done

SWvheerden committed:
code rabbit

SWvheerden committed:
fixes

SWvheerden committed:
fix cw update
